tools/swift-reflection-dump/swift-reflection-dump.cpp

```diff
@@ -120,6 +137,93 @@ static bool needToRelocate(SectionRef S) {
   return true;
 }
 
+static section_iterator findSectionByName(StringRef Name, const ObjectFile &O) {
+  auto Sections = O.sections();
+  SmallVector<section_iterator, 1> FoundSections;
+  std::copy_if(Sections.begin(), Sections.end(),
+               std::back_inserter(FoundSections), [Name](SectionRef S) {
+                 StringRef N;
+                 if (auto EC = S.getName(N))
+                   reportError(EC);
+                 return N == Name;
+               });
+  if (FoundSections.size() != 1)
+    return Sections.end();
+  return FoundSections.front();
+}
+
+template <class ELFObj,
+          typename std::enable_if<
+              std::is_base_of<ELFObjectFileBase, ELFObj>::value, int>::type = 0>
+const typename ELFObj::Elf_Rela *findRelaForOffset(uint64_t Offset,
+                                                   const ELFObj &O) {
+  for (const SectionRef &S :
+       static_cast<const ObjectFile *>(&O)->dynamic_relocation_sections()) {
+    const auto *Sec = O.getSection(S.getRawDataRefImpl());
+    if (Sec->sh_type != llvm::ELF::SHT_RELA)
+      continue;
+    for (const RelocationRef &R : S.relocations()) {
+      const auto *Rela = O.getRela(R.getRawDataRefImpl());
+      if (Rela->r_offset == Offset &&
+          Rela->getType(false /* isMips64EL */) == llvm::ELF::R_X86_64_RELATIVE)
+        return Rela;
+    }
+  }
+  return nullptr;
+}
+
+static void updateReflectionMetdataNote(const ObjectFile &O, char *Memory) {
+#if !(defined(_WIN32) || defined(__APPLE__) && defined(__MACH__))
```

Comment:
Why not make this `#if defined(__ELF__)`?

Reply:
Maybe this is a larger refactoring, but it seems like this ought to be based on the ObjectFile's format rather than a compile-time #define chosen by the host.

Reply:
yeah, and almost everything (sadly not everything, but I'd prefer to address those issues in a separate patch) would work here, in this particular file I had to use #ifdefs to include ImageInspectionELF.h to get the macro SWIFT_REFLECTION_METADATA_ELF_NOTE_MAGIC_STRING which is available only on ELF platforms. I'm kinda not sure what the best way to do it - so we can either redefine it here / create a named constant or move that macro into a different header which is available on all the platforms (ImageInspectionELF.h is not - it starts with #if defined(__ELF__) .. )

Reply:
I'm somehow hesitating to put this ELF-specific macro into ImageInspection.h, probably would be better to simply redefine it inside swift-reflection-dump.cpp

Reply:
I'm not sure about moving that to `swift-reflection-dump.cpp`, as that's not the only user of this.

Reply:
yeah, I agree, need to think

Reply:
so now this macro (SWIFT_REFLECTION_METADATA_ELF_NOTE_MAGIC_STRING) is defined in ImageInspectionELF.h but not guarded by 
#if defined(__ELF__).